api/turing/imagebuilder/imagebuilder.go

```diff
@@ -0,0 +1,311 @@
+package imagebuilder
+
+import (
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/google/go-containerregistry/pkg/authn"
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/google/go-containerregistry/pkg/v1/google"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"github.com/google/go-containerregistry/pkg/v1/remote/transport"
+	"github.com/pkg/errors"
+	apibatchv1 "k8s.io/api/batch/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+
+	"github.com/gojek/turing/api/turing/cluster"
+	"github.com/gojek/turing/api/turing/config"
+	"github.com/gojek/turing/api/turing/log"
+	"github.com/gojek/turing/api/turing/models"
+)
+
+// JobStatus is the current state of the image building job.
+type JobStatus string
+
+const (
+	tickDurationInSeconds = 5
+	// JobStatusActive is the status of the image building job is active
+	JobStatusActive = JobStatus(iota)
+	// JobStatusFailed is when the image building job has failed
+	JobStatusFailed
+	// JobStatusSucceeded is when the image building job has succeeded
+	JobStatusSucceeded
+	// JobStatusUnknown is when the image building job status is unknown
+	JobStatusUnknown
+)
+
+// BuildImageRequest contains the information needed to build the OCI image
+type BuildImageRequest struct {
+	ProjectName string
+	ModelName   string
```

Comment:
Hmm? Why ResourceName? We are always building machine learning models right?

Reply:
Isn't it ensembler (and not a model)?

Reply:
> We are always building machine learning models right?

Here it is ensembler, right?

> Why ResourceName?

Imagebuilder is a generic service and it can be used to build any Docker image. Currently it is being used to build ensembler image. IDK if we will use it for something else in the future but it seems it doesn't need to be tied to a Turing-specific concept. That's why I proposed ResourceName.

Reply:
Well, I wouldn't say any docker image because the flags/Dockerfile are kind of hardcoded. https://github.com/ashwinath/turing/blob/batch-ensembler-orchestration/engines/batch-ensembler/app.Dockerfile#L6

It can only be used to build models.

Reply:
Ehh the model url arg is only used to do `RUN gsutil -m cp -r ${MODEL_URL} .` right? I don't imagine it is playing a role anywhere else (because it is using ARG and not ENV).

And you are setting it in the Dockerfile and [here](https://github.com/ashwinath/turing/blob/batch-ensembler-orchestration/api/turing/imagebuilder/imagebuilder.go#L211) so that can be freely changed. Please CMIIW.

Reply:
This is just to download the model and bake it into the container. The model is set to this destination: https://github.com/ashwinath/turing/blob/batch-ensembler-orchestration/api/turing/service/ensembling_job_service.go#L129-L139 and used here if I'm not wrong: https://github.com/ashwinath/turing/blob/batch-ensembler-orchestration/engines/batch-ensembler/ensembler/ensembler.py#L33

Reply:
But I just realised that the Dockerfile can be changed anyway, so actually anything can be changed, but I'm not sure how it's going to be changed.

Reply:
Will change this to ResourceName I'm convinced now